pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/RealtimeSegmentAssignment.java

```diff
@@ -172,6 +183,39 @@ private void checkReplication(InstancePartitions instancePartitions) {
     }
   }
 
+  private int getSegmentPartitionId(String segmentName) {
+    // A fast path if the segmentName is a LLC segment name and we can get the partition id from the name directly.
+    if (LLCSegmentName.isLowLevelConsumerSegmentName(segmentName)) {
+      return new LLCSegmentName(segmentName).getPartitionGroupId();
+    }
+    // Otherwise, retrieve the partition id from the segment zk metadata. Currently only realtime segments from upsert
+    // enabled tables have partition ids in their segment metadata.
+    RealtimeSegmentZKMetadata segmentZKMetadata = ZKMetadataProvider
+        .getRealtimeSegmentZKMetadata(_helixManager.getHelixPropertyStore(), _realtimeTableName, segmentName);
+    Preconditions
+        .checkState(segmentZKMetadata != null, "Failed to find segment ZK metadata for segment: %s of table: %s",
+            segmentName, _realtimeTableName);
+    return getSegmentPartitionIdFromZkMetaData(segmentZKMetadata);
+  }
+
+  private int getSegmentPartitionIdFromZkMetaData(RealtimeSegmentZKMetadata segmentZKMetadata) {
+    String segmentName = segmentZKMetadata.getSegmentName();
+    Preconditions.checkState(segmentZKMetadata.getPartitionMetadata() != null,
+        "Segment ZK metadata for segment: %s of table: %s does not contain partition metadata", segmentName,
+        _realtimeTableName);
+
+    ColumnPartitionMetadata partitionMetadata =
+        segmentZKMetadata.getPartitionMetadata().getColumnPartitionMap().get(_partitionColumn);
+    Preconditions.checkState(partitionMetadata != null,
+        "Segment ZK metadata for segment: %s of table: %s does not contain partition metadata for column: %s",
```

Comment:
shall we make the message more informative, by adding `checking if the table is in upsert mode`?

Reply:
good idea. done.